Migrate Block Patterns Registration to jetpack-mu-wpcom

okmttdhr opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->

This PR migrates the Block Patterns feature (inserting wpcom Block Patterns to the editor) into `jetpack-mu-wpcom`, which was originally located in ETK ([editing-toolkit-plugin/block-patterns](https://github.com/Automattic/wp-calypso/tree/e4467c3/apps/editing-toolkit/editing-toolkit-plugin/block-patterns)). 

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->



## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Patch this PR by following the instructions https://github.com/Automattic/jetpack/pull/34162#issuecomment-1814111621
* Disable ETK 
	* on Simple site: By removing this filter https://github.com/Automattic/wp-calypso/blob/b81dab43af076713053372cf2ca8f32fb5a90a3d/apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php#L286-L296
	* on Atomic site: By uninstalling ETK
* Go to the Post Editor
* Open the block inserter
* Check for categories (e.g., Intro)
* Try inserting patterns

<img width="646" alt="Screenshot 2023-11-16 at 12 16 13" src="https://github.com/Automattic/jetpack/assets/5287479/687dbc58-c05c-468e-8876-19cd4c781cb9">

okmttdhr committed:
Migrate Block Patterns

okmttdhr committed:
Add changelog

okmttdhr committed:
Bump the version

okmttdhr committed:
Add the prefix to the class name

okmttdhr committed:
Make it compatible with older PHP versions

okmttdhr committed:
Do nothing if it is loaded in the ETK

okmttdhr committed:
Move Block Patterns into load_features

obenland commented on projects/packages/jetpack-mu-wpcom/src/features/block-patterns/class-wpcom-block-patterns-utils.php:
```diff
@@ -0,0 +1,149 @@
+	public function get_block_patterns_locale() {
+		// Block patterns display in the user locale.
+		$language = get_user_locale();
+		return \A8C\FSE\Common\get_iso_639_locale( $language );
```
Will we need to migrate this function, too?

okmttdhr replied:
I changed to use `Common\get_iso_639_locale` that was already defined in `jetpack-mu-wpcom/src/common/index.php`!

obenland commented on projects/packages/jetpack-mu-wpcom/src/features/block-patterns/class-wpcom-block-patterns-utils.php:
```diff
@@ -0,0 +1,149 @@
+		// the template related types and to pages - this is to avoid the pattern appearing in
+		// the inserter for posts and other post types. Pages are included because it's not unusual
+		// to use a blank template and add a specific header and footer to a page.
+		if ( $block_types_count && count(
+			array_filter(
+				$block_types,
+				function ( $block_type ) {
+					return preg_match( '#core/template-part/#', $block_type );
+				}
+			)
+		) === $block_types_count ) {
```
This is really hard to read. Could we pull out the `count( array_filter() )` and save that in a variable and compare that?

obenland commented on projects/packages/jetpack-mu-wpcom/src/features/block-patterns/load-block-patterns.php:
```diff
@@ -0,0 +1,77 @@
+ * Re-register some core patterns to push them down in the inserter list.
+ * The reason is that Dotcom curate the pattern list based on their look.
+ */
+function reorder_curated_core_patterns() {
```
We should add function prefixes in this file

okmttdhr committed:
Add wpcom prefix

okmttdhr committed:
Use Jetpack_Mu_Wpcom\Common

okmttdhr committed:
Create $template_parts variable

okmttdhr committed:
Merge branch 'trunk' into add/migrate-block-patterns

anomiex commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -41,6 +41,7 @@ public static function init() {
 
 		add_action( 'plugins_loaded', array( __CLASS__, 'load_first_posts_stream_helpers' ) );
 		add_action( 'plugins_loaded', array( __CLASS__, 'load_domain_email_nag' ) );
+		add_action( 'plugins_loaded', array( __CLASS__, 'load_block_patterns' ) );
```
This function does not exist, which is what's causing tests on D129217-code to fail.

anomiex commented on projects/packages/jetpack-mu-wpcom/src/features/block-patterns/test/class-wpcom-block-patterns-from-api-test.php:
These tests are not being run. PHPUnit tests would normally be placed in `projects/packages/jetpack-mu-wpcom/tests/php/features/block-patterns/` rather than `projects/packages/jetpack-mu-wpcom/src/features/block-patterns/test/`.

miksansegundo committed:
Fix tests

miksansegundo committed:
Fix test

miksansegundo committed:
Use package version rather than ETK version for patterns cache

miksansegundo commented:
@okmttdhr, I have created a new PR https://github.com/Automattic/jetpack/pull/34249, feel free to merge it into this PR.

The new PR works on simple and atomic, passes all checks, addresses the feedback in this PR, and fixes some other things. cc: @obenland @anomiex

okmttdhr committed:
Merge branch 'trunk' into add/migrate-block-patterns

okmttdhr committed:
Remove load- prefix

okmttdhr committed:
Add prefix to tests

okmttdhr committed:
block-patterns feature should be loaded in load_features

okmttdhr committed:
Block_Patterns_Utils -> Wpcom_Block_Patterns_Utils

okmttdhr commented:
> @okmttdhr, I have created a new PR #34249, feel free to merge it into this PR.
> 
> The new PR works on simple and atomic, passes all checks, addresses the feedback in this PR, and fixes some other things. cc: @obenland @anomiex

I've cherry-picked the commits (as I wanted to keep the prefix and screwed up the diff with merge). Thanks for working on that! 😃

miksansegundo reviewed (approved):
I have tested that it works on simple and atomic sites 🎉. cc: @Automattic/lego 

**1. Simple sites:**

  - Test 1.1 - Apply this PR in my sandbox to verify that Dotcom patterns are registered from the ETK plugin
  - Test 1.2 - Apply this PR in my sandbox together with the PR https://github.com/Automattic/wp-calypso/pull/84256 to verify that Dotcom patterns are registered from jetpack-mu-wpcom

**2. Atomic sites:**

  - Test 2.1 - Apply this PR on my site via Jetpack beta plugin to verify that Dotcom patterns are registered from the ETK plugin
  - Test 2.2 - Apply this PR and deactivating the ETK plugin to verify that Dotcom patterns are registered from jetpack-mu-wpcom
- Test 2.3 - Apply this PR and deleting the ETK plugin to verify that Dotcom patterns are registered from jetpack-mu-wpcom
